### DISCUSSION TITLE
Load nodeset xml into server

### DISCUSSION THREAD

**Discussion by pfetrial** [FreeOpcUa/python-opcua#1202] [General]
hello everyone.
I wish you a happy new year!
I have a question concerning the python server : Is it possible to load a information model  ( Xml) into a server ? If yes how ? I haven't seen some functions related to this .. thanks

**Comment by AndreasHeine** [comment]
https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/opcua/server/server.py#L513

https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/examples/server-import-robot-nodeset.py#L15

**Reply by xenonforlife** [reply]
Hi @AndreasHeine, @swamper123, @oroulet  I am trying to do the same as OP.  I cloned the repo on Windows 10 and the test server runs fine. However when I am trying to run the server-import-robot-nodeset.py file, I am always getting a "No such file or directory error". I made sure that the appropriate nodesets are all available in the schemas folder and further within the appropriate folders e.g. DI and Robotics. I still cannot make python find the xml files somehow. I also tried using Path from the pathlib library to resolve any path problems. However it still doesn't work. The only way I could make it work was using absolute paths, which unfortunately is not the solution I am looking for.  Even if I copy the nodeset files into the same examples folder (as the server-import-robot-nodeset.py file) and change the argument of `server.import` to just `"Opc.Ua.Di.NodeSet2.xml"`, it still does not find the file. Could you please help me here?

**Reply by AndreasHeine** [reply]
os.path() module ?

i used this alot:
```python
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
server.import_xml(os.path.join(BASE_DIR, "nodeset", "Opc.Ua.Di.NodeSet2.xml"))
```

**Reply by xenonforlife** [reply]
Haven't tried that since I am on Python 3.7 so the recommended way is pathlib.Path. Anyway I will give it a shot and report back. Thanks for your super quick reply!

**Reply by xenonforlife** [reply]
@AndreasHeine Worked like a charm, thank you so much!

**Comment by pfetrial** [comment]
Oh thanks . I haven’t seen  this API.

 

 

Von: Andreas Heine <notifications@github.com> 
Gesendet: Freitag, 1. Januar 2021 10:30
An: FreeOpcUa/python-opcua <python-opcua@noreply.github.com>
Cc: pfetrial <patrick.frey37@sfr.fr>; Author <author@noreply.github.com>
Betreff: Re: [FreeOpcUa/python-opcua] Load nodeset xml into server (#1202)

 

https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/opcua/server/server.py#L513

**Comment by pfetrial** [comment]
Hello Everyone,
I succeded to import a xml file into the server. Some xml file are not imported , iat this moment I can't figure out what are the criterias to succed or not, but  perhaps It depends on the complexity of the file, size of the file?...
I have seen that the flag "ishistorizing" is not imported . ( it is always false in the server)
Is the possibility to have a callback function when the import function adds an item in the tree?
Thx

**Reply by swamper123** [reply]
Some Nodes are not imported or XML-Files are not imported?

**Reply by AndreasHeine** [reply]
"ishistorizing" thats right, it is set if you add historizing for thtat node its not part of the xml import!
look in the historizing examples to see how it works!
https://github.com/FreeOpcUa/python-opcua/blob/1cca33dd5fb70858b5e811b08356ca97520bcbaf/examples/server-datavalue-history.py#L36

**Comment by oroulet** [comment]
the list of imported nodeid is returned by the import_xml() method and hoepfully a log is written when  node is not imported (not sure about the latest)

**Reply by swamper123** [reply]
Nodes are returned, problematic references and nodes are logged as warning in that method.
